server/v2/stf/stf.go

```diff
@@ -108,17 +104,48 @@ func (s STF[T]) validateTx(ctx context.Context, state store.WritableState, gasLi
 	return validateCtx.gasUsed, validateCtx.events, applyStateChanges(state, validateState)
 }
 
+// execTx executes the tx messages on the provided state. If the tx fails then the state is discarded.
 func (s STF[T]) execTx(ctx context.Context, state store.WritableState, gasLimit uint64, tx T) (msgResp Type, gasUsed uint64, execEvents []event.Event, err error) {
 	execState := s.branch(state)
 	execCtx := s.makeContext(ctx, tx.GetSenders(), execState, gasLimit)
 	// atomic execution of the all messages in a transaction, TODO: we should allow messages to fail in a specific mode
-	for _, msg := range tx.GetMessages() {
-		msgResp, err = s.handleMsg(execCtx, msg)
-		if err != nil {
-			return nil, 0, nil, err
+	var txErr error
+	for i, msg := range tx.GetMessages() {
+		msgResp, txErr = s.handleMsg(execCtx, msg)
+		if txErr != nil {
+			txErr = fmt.Errorf("tx execution failed at message with index %d: %w", i, txErr)
+			break // stop execution when one message fails.
+		}
+	}
+	// if tx failed then we run the post tx handler with the initial state, and we only return
+	// post tx handler events.
+	if txErr != nil {
+		postTxEvents, postTxErr := s.runPostTxHandler(ctx, state, tx, false)
+		if postTxErr != nil {
+			return nil, execCtx.gasUsed, nil, errors.Join(txErr, postTxErr)
 		}
+		return nil, execCtx.gasUsed, postTxEvents, txErr
+	}
+	// if the tx did not fail, we run the post handler using a branch of a branch of the provided
+	// initial state. Rationale for running in a branch of a branch is that in case the post tx fails
+	// then the whole exec state needs to be rolled back.
```

Comment:
if the post handler fails and we roll back the whole thing tx exec, then why branch again?

Reply:
also, could we do something like:
```
postTxEvents, postTxErr := s.runPostTxHandler(ctx, state, tx, txErr == nil)
```
`errors.Join` will discard any passed in errors that are nil: https://pkg.go.dev/errors#Join

Reply:
solved!